docs/cloud-init.md

```diff
@@ -65,6 +65,62 @@ spec:
         os: hvm
     type: qemu
 ```
+
+### NoCloud with UserData stored in k8s secret
+
+Instead of placing the userdata directly into the VM definition, another option
+is to reference the userdata in a k8s secret object.
+
+Example: Create a k8s secret with UserData. Reference the secret in the NoCloud
+disk definition.
+
+First create the secret.
+```
+apiVersion: v1
+kind: Secret
+metadata:
+  name: my-vm-secret
+type: Opaque
+data:
+  userdata: I2Nsb3VkLWNvbmZpZwpwYXNzd29yZDogYXRvbWljCnNzaF9wd2F1dGg6IFRydWUKY2hwYXNzd2Q6IHsgZXhwaXJlOiBGYWxzZSB9Cg==
+```
+
+Then reference the secret in the userDataSecretRef field.
+```
+metadata:
+  name: testvm-nocloud
+apiVersion: kubevirt.io/v1alpha1
+kind: VM
+spec:
+  domain:
+    devices:
+      disks:
+      - type: ContainerRegistryDisk:v1alpha
+        source:
+          name: kubevirt/cirros-registry-disk-demo:devel
+        target:
+          dev: vda
+      - type: file
+        target:
+          dev: vdb
+        cloudinit:
+            nocloud:
+                userDataSecretRef: my-vm-userdata 
```

Comment:
Just an observation (It may or may not be helpful in this context): it is possible to also mount a secret into a pod.

```yaml
  volumeMounts:
  - name: cloud-user-data
    mountPath: .....
    readOnly: true

  volumes:
    - name: cloud-user-data
      secret:
        secretName: my-vm-userdata
```

This might not be helpful if libvirt needs access to this data before the image has been booted--and might impact migrations.

Reply:
yeah, assigning the secret to a pod would only help us if we were exposing the secret to the virt-launcher component.  virt-launcher isn't involved with the iso creation anymore though.

Instead, virt-handler pulls the secret and uses that secret when it generates the NoCloud iso.